File: src/tls/parser.py
"""
Pure-Python TLS ClientHello Parser.
High-performance, zero-dependency parser for JA4 fingerprinting.
Eliminates Scapy IPC overhead.
"""

from typing import Dict, List, Optional
import struct

def parse_client_hello(data: bytes) -> Optional[Dict]:
    """
    Parse a raw TLS ClientHello record.
    Returns a dict compatible with Scapy's TLSParser output or None on failure.
    Memory: O(N) linear pass. Never raises.
    """
    try:
        # 1. TLS Record Header (5 bytes)
        # 0: Content Type (0x16 = Handshake)
        # 1-2: Version
        # 3-4: Length
        if len(data) < 5 or data[0] != 0x16:
            return None
        
        record_len = struct.unpack("!H", data[3:5])[0]
        # Allow for data being larger than one record (peek buffer)
        # but the record itself must be complete in the buffer.
        if len(data) < 5 + record_len:
            return None
        
        # 2. Handshake Header (4 bytes)
        # 0: Handshake Type (0x01 = ClientHello)
        # 1-3: Length
        pos = 5
        if data[pos] != 0x01:
            return None
        
        hs_len = struct.unpack("!I", b"\x00" + data[pos+1:pos+4])[0]
        if record_len < 4 + hs_len:
            return None
        
        pos += 4
        
        # 3. ClientHello Body
        # 0-1: Version (Legacy)
        # 2-33: Random (32 bytes)
        # 34: Session ID Length
        if len(data) < pos + 35:
            return None
        
        legacy_version = struct.unpack("!H", data[pos:pos+2])[0]
        pos += 34 # Version + Random
        
        session_id_len = data[pos]
        pos += 1 + session_id_len
        
        # 4. Cipher Suites
        if len(data) < pos + 2:
            return None
        cs_len = struct.unpack("!H", data[pos:pos+2])[0]
        pos += 2
        
        if len(data) < pos + cs_len:
            return None
        
        cipher_suites = []
        for i in range(0, cs_len, 2):
            cipher_suites.append(struct.unpack("!H", data[pos+i:pos+i+2])[0])
        pos += cs_len
        
        # 5. Compression Methods
        if len(data) < pos + 1:
            return None
        cm_len = data[pos]
        pos += 1 + cm_len
        
        # 6. Extensions
        fields = {
            "version": legacy_version,
            "cipher_suites": cipher_suites,
            "extensions": [],
            "sni": "",
            "alpn": [],
            "supported_versions": [],
            "supported_groups": [],
            "signature_algorithms": [],
        }
        
        if len(data) < pos + 2:
            # No extensions is valid in very old TLS
            return fields
            
        exts_total_len = struct.unpack("!H", data[pos:pos+2])[0]
        pos += 2
        
        exts_end = pos + exts_total_len
        if len(data) < exts_end:
            return fields # Truncated extensions? Fail open.
            
        while pos + 4 <= exts_end:
            ext_type = struct.unpack("!H", data[pos:pos+2])[0]
            ext_len = struct.unpack("!H", data[pos+2:pos+4])[0]
            pos += 4
            
            if pos + ext_len > exts_end:
                break
                
            fields["extensions"].append(ext_type)
            ext_data = data[pos:pos+ext_len]
            
            # Parse specific extensions needed for JA4
            if ext_type == 0: # server_name (SNI)
                try:
                    # 0-1: List Length
                    # 2: Name Type (0 = host_name)
                    # 3-4: Name Length
                    if len(ext_data) >= 5 and ext_data[2] == 0:
                        name_len = struct.unpack("!H", ext_data[3:5])[0]
                        if len(ext_data) >= 5 + name_len:
                            fields["sni"] = ext_data[5:5+name_len].decode("ascii", errors="ignore")
                except Exception:
                    pass
            elif ext_type == 16: # ALPN
                try:
                    # 0-1: List Length
                    if len(ext_data) >= 3:
                        alpn_len = struct.unpack("!H", ext_data[0:2])[0]
                        alpn_pos = 2
                        while alpn_pos < 2 + alpn_len:
                            p_len = ext_data[alpn_pos]
                            alpn_pos += 1
                            if alpn_pos + p_len <= len(ext_data):
                                fields["alpn"].append(
                                    ext_data[alpn_pos:alpn_pos+p_len].decode("ascii", errors="ignore")
                                )
                            alpn_pos += p_len
                except Exception:
                    pass
            elif ext_type == 43: # supported_versions
                try:
                    # 0: List Length
                    if len(ext_data) >= 3:
                        v_len = ext_data[0]
                        for i in range(1, v_len, 2):
                            fields["supported_versions"].append(
                                struct.unpack("!H", ext_data[i:i+2])[0]
                            )
                except Exception:
                    pass
            elif ext_type == 10: # supported_groups
                try:
                    if len(ext_data) >= 4:
                        g_len = struct.unpack("!H", ext_data[0:2])[0]
                        for i in range(2, 2 + g_len, 2):
                            fields["supported_groups"].append(
                                struct.unpack("!H", ext_data[i:i+2])[0]
                            )
                except Exception:
                    pass
            elif ext_type == 13: # signature_algorithms
                try:
                    if len(ext_data) >= 4:
                        s_len = struct.unpack("!H", ext_data[0:2])[0]
                        for i in range(2, 2 + s_len, 2):
                            fields["signature_algorithms"].append(
                                struct.unpack("!H", ext_data[i:i+2])[0]
                            )
                except Exception:
                    pass
            
            pos += ext_len
            
        return fields

    except Exception:
        return None

File: internal/tls/parser.go
package tls

import (
	"unsafe"
	"encoding/binary"
	"errors"
	"fmt"
)

// Minimum valid ClientHello size:
// TLS record header (5) + handshake header (4) + legacy_version (2) +
// random (32) + session_id_len (1) = 44 bytes minimum before cipher suites.
const minClientHelloSize = 44

// ErrNotTLS is returned when the input does not look like a TLS record.
var ErrNotTLS = errors.New("not a TLS record")

// ErrNotClientHello is returned when the handshake type is not ClientHello.
var ErrNotClientHello = errors.New("not a ClientHello message")

// ErrTruncated is returned when the record is too short to parse.
var ErrTruncated = errors.New("truncated ClientHello")

// ParseClientHello parses raw bytes containing a TLS ClientHello record
// and returns the structured ClientHelloInfo.
//
// It never panics on malformed input; all errors are returned as values.
// The caller should treat any error as "not a recognisable ClientHello" and
// fail open.
func ParseClientHello(data []byte) (*ClientHelloInfo, error) {
	if len(data) < 5 {
		return nil, ErrTruncated
	}

	// TLS Record Layer:
	//   byte 0:    content type (0x16 = Handshake)
	//   bytes 1-2: legacy record version
	//   bytes 3-4: record length (uint16 big-endian)
	if data[0] != 0x16 {
		return nil, ErrNotTLS
	}
	recordLen := int(binary.BigEndian.Uint16(data[3:5]))
	if recordLen > 16384 {
		return nil, errors.New("TLS record too large")
	}
	if len(data) < 5+recordLen {
		return nil, ErrTruncated
	}

	// Handshake Header (within record body):
	//   byte 0:    handshake type (0x01 = ClientHello)
	//   bytes 1-3: body length (uint24 big-endian)
	body := data[5 : 5+recordLen]
	if len(body) < 4 {
		return nil, ErrTruncated
	}
	if body[0] != 0x01 {
		return nil, ErrNotClientHello
	}

	helloLen := int(body[1])<<16 | int(body[2])<<8 | int(body[3])
	if len(body) < 4+helloLen {
		return nil, ErrTruncated
	}
	if 4+helloLen < minClientHelloSize {
		return nil, ErrTruncated
	}

	hello := body[4 : 4+helloLen]

	info := &ClientHelloInfo{}
	info.Raw = data

	// ClientHello body:
	//   bytes 0-1:  legacy_version
	//   bytes 2-33: random (32 bytes)
	//   byte 34:    session_id length
	//   bytes 35..34+sid_len: session_id
	if len(hello) < 34 {
		return nil, ErrTruncated
	}
	info.LegacyVersion = binary.BigEndian.Uint16(hello[0:2])

	pos := 2 + 32 // skip version + random
	if pos >= len(hello) {
		return nil, ErrTruncated
	}
	sidLen := int(hello[pos])
	pos++
	pos += sidLen // skip session ID

	// Cipher suites: uint16 length + pairs of uint16
	if pos+2 > len(hello) {
		return nil, ErrTruncated
	}
	csLen := int(binary.BigEndian.Uint16(hello[pos : pos+2]))
	pos += 2
	if pos+csLen > len(hello) {
		return nil, ErrTruncated
	}
	if csLen%2 != 0 {
		return nil, fmt.Errorf("invalid cipher suites length %d", csLen)
	}
	for i := 0; i < csLen; i += 2 {
		cs := binary.BigEndian.Uint16(hello[pos+i : pos+i+2])
		if info.CipherSuites == nil { info.CipherSuites = make([]uint16, 0, csLen/2) }
		info.CipherSuites = append(info.CipherSuites, cs)
	}
	pos += csLen

	// Compression methods: uint8 length + bytes
	if pos+1 > len(hello) {
		return nil, ErrTruncated
	}
	compLen := int(hello[pos])
	pos++
	if pos+compLen > len(hello) {
		return nil, ErrTruncated
	}
	info.CompressionMethods = hello[pos : pos+compLen]
	pos += compLen

	// Extensions (optional — may be absent for very old clients)
	if pos >= len(hello) {
		return info, nil
	}
	if pos+2 > len(hello) {
		return nil, ErrTruncated
	}
	extTotalLen := int(binary.BigEndian.Uint16(hello[pos : pos+2]))
	pos += 2
	if pos+extTotalLen > len(hello) {
		return nil, ErrTruncated
	}

	extData := hello[pos : pos+extTotalLen]
	if err := parseExtensions(info, extData); err != nil {
		return nil, err
	}

	return info, nil
}

// parseExtensions iterates the extensions block and populates info.
func parseExtensions(info *ClientHelloInfo, data []byte) error {
	pos := 0
	for pos < len(data) {
		if pos+4 > len(data) {
			return ErrTruncated
		}
		extType := binary.BigEndian.Uint16(data[pos : pos+2])
		extLen := int(binary.BigEndian.Uint16(data[pos+2 : pos+4]))
		pos += 4
		if pos+extLen > len(data) {
			return ErrTruncated
		}
		extBody := data[pos : pos+extLen]
		pos += extLen

		if info.Extensions == nil { info.Extensions = make([]uint16, 0, 16) }
		info.Extensions = append(info.Extensions, extType)

		switch extType {
		case 0x0000: // SNI
			info.SNIPresent = true
			parseSNI(info, extBody)
		case 0x0010: // ALPN
			parseALPN(info, extBody)
		case 0x002b: // SupportedVersions
			parseSupportedVersions(info, extBody)
		case 0x000a: // SupportedGroups
			parseSupportedGroups(info, extBody)
		case 0x000d: // SignatureAlgorithms
			parseSignatureAlgorithms(info, extBody)
		}
	}
	return nil
}

// parseSNI extracts the server name from the SNI extension body.
// SNI body format:
//
//	server_name_list_length (uint16) + entries:
//	  name_type (uint8, 0=host_name) + name_length (uint16) + name_bytes
func parseSNI(info *ClientHelloInfo, data []byte) {
	if len(data) < 5 {
		return
	}
	listLen := int(binary.BigEndian.Uint16(data[0:2]))
	if listLen < 3 || 2+listLen > len(data) {
		return
	}
	// Only the first server name entry is used (TLS SNI carries one name in practice).
	p := 2
	if p+3 > len(data) {
		return
	}
	// nameType := data[p] // 0 = host_name
	nameLen := int(binary.BigEndian.Uint16(data[p+1 : p+3]))
	p += 3
	if p+nameLen > len(data) {
		return
	}
	info.SNI = unsafeString(data[p : p+nameLen])
}

// parseALPN extracts protocol names from the ALPN extension body.
// ALPN body format:
//
//	protocol_name_list_length (uint16) + entries:
//	  name_length (uint8) + name_bytes
func parseALPN(info *ClientHelloInfo, data []byte) {
	if len(data) < 2 {
		return
	}
	listLen := int(binary.BigEndian.Uint16(data[0:2]))
	if 2+listLen > len(data) {
		return
	}
	p := 2
	end := 2 + listLen
	for p < end {
		if p+1 > end {
			return
		}
		nameLen := int(data[p]) //nolint:gosec // bounds checked: p < end <= len(data)
		p++
		if p+nameLen > end {
			return
		}
		info.ALPNProtocols = append(info.ALPNProtocols, unsafeString(data[p:p+nameLen]))
		p += nameLen
	}
}

// parseSupportedVersions extracts TLS versions from the supported_versions extension.
// Body format (ClientHello): list_length (uint8) + uint16 versions
func parseSupportedVersions(info *ClientHelloInfo, data []byte) {
	if len(data) < 1 {
		return
	}
	listLen := int(data[0])
	if 1+listLen > len(data) || listLen%2 != 0 {
		return
	}
	for i := 0; i < listLen; i += 2 {
		v := binary.BigEndian.Uint16(data[1+i : 3+i])
		info.SupportedVersions = append(info.SupportedVersions, v)
	}
}

// parseSupportedGroups extracts named groups from extension 0x000a.
func parseSupportedGroups(info *ClientHelloInfo, data []byte) {
	if len(data) < 2 {
		return
	}
	listLen := int(binary.BigEndian.Uint16(data[0:2]))
	if 2+listLen > len(data) || listLen%2 != 0 {
		return
	}
	for i := 0; i < listLen; i += 2 {
		g := binary.BigEndian.Uint16(data[2+i : 4+i])
		info.SupportedGroups = append(info.SupportedGroups, g)
	}
}

// parseSignatureAlgorithms extracts signature algorithms from extension 0x000d.
func parseSignatureAlgorithms(info *ClientHelloInfo, data []byte) {
	if len(data) < 2 {
		return
	}
	listLen := int(binary.BigEndian.Uint16(data[0:2]))
	if 2+listLen > len(data) || listLen%2 != 0 {
		return
	}
	for i := 0; i < listLen; i += 2 {
		s := binary.BigEndian.Uint16(data[2+i : 4+i])
		info.SignatureAlgorithms = append(info.SignatureAlgorithms, s)
	}
}

func unsafeString(b []byte) string {
	return unsafe.String(unsafe.SliceData(b), len(b))
}
